Aggregate mozilla-central script

zbraniecki opened the pull request:
This is the script I'm using to calculate progress of mozilla-central migration to Fluent.

zbraniecki commented:
> Concretely, if nobody but you can run the thing, it's an r-.

I don't understand. What do you need to be able to run the thing except of adjusting the two variables at the top?

> Looking at irc and this patch, the revset is wrong, I still think the README has the right one.

The revset in README cannot work since it uses `pushhead("central")` and pushhead takes no arguments - https://hg.mozilla.org/hgcustom/version-control-tools/file/tip/hgext/pushlog/__init__.py#l880

I tested the revset locally and it seems to work. Can you be more specific on what makes the revset incorrect (and how to correct it if you know :)).

> And I prefer --cwd over positional params.

On it.

Pike commented:
> The revset in README cannot work since it uses `pushhead("central")` and pushhead takes no arguments - https://hg.mozilla.org/hgcustom/version-control-tools/file/tip/hgext/pushlog/__init__.py#l880

The one in mozext does, https://dxr.mozilla.org/hgcustom_version-control-tools/rev/307f3c28687630bf91b4c19c913f0c677e0ae724/hgext/mozext/__init__.py#54-55

Here's what I have in `.hg/hgrc` in m-c:

```
[extensions]
fsmonitor =
mozext = ~/.mozbuild/version-control-tools/hgext/mozext
firefoxtree = ~/.mozbuild/version-control-tools/hgext/firefoxtree
push-to-try = ~/.mozbuild/version-control-tools/hgext/push-to-try
```

zbraniecki commented:
> The one in mozext does, https://dxr.mozilla.org/hgcustom_version-control-tools/rev/307f3c28687630bf91b4c19c913f0c677e0ae724/hgext/mozext/__init__.py#54-55

As per `sheehan` who maintains the vct, `mozext` is deprecated:

> sheehan> gandalf: can you try disabling `mozext` and running that command again? the mozext implementation uses the "changetracker.db" database, whereas `pushlog` uses `pushlog2.db`. I have plans to remove the former

zbraniecki commented:
Updated per Pike's verbal feedback:

 - Switched to command line arguments for mc path and gh-pages path
 - Switched the query to use the last push before the date, rather than the first after
 - Removed `merge()` from revset

Read for review

Pike commented on aggregate-mc.py:
```diff
@@ -0,0 +1,128 @@
+    pick_next_revision, get_current_revision, get_revision_date, \
+    switch_to_revision, write_data, is_file_writable, is_dir_readable
+
+DATA_PATH = "./data"
```
Can we just use this once? Seems that we only ever use gh_pages + data, we could do that when handling the command args.

Pike commented on aggregate-mc.py:
```diff
@@ -0,0 +1,128 @@
+            if ext and ext not in progress:
+                progress[ext] = count
+            else:
+                progress[ext] += count
```
Use a `defaultdict(int)` here? Less hassle.

Pike commented on aggregate-mc.py:
```diff
@@ -0,0 +1,128 @@
+        for path, count in subset.items():
+            entries.append({"file": path, "count": count})
+            ext = os.path.splitext(path)[1]
+            if not ext:
```
Do we actually hit this? If you want to keep it, move it above the snapshot part? It's a bit weird to me that we handle the two differently.

zbraniecki replied:
I don't understand. I'm splitting a path and if the path doesn't contain an extension (for example the file is `foo` instead of `foo.txt`), I wanted to gracefully skip such a file. Where do you want me to move that?

Pike replied:
before the `entries.append`, as that's unconditional.

Pike commented on aggregate-mc.py:
```diff
@@ -0,0 +1,128 @@
+    next_revision = pick_next_revision(last_date, START_DATE, mc_path,
+                                       FREQUENCY)
+
+    while next_revision:
```
Make this `while True:` ? We never hit this exit.

Pike commented on aggregate-mc.py:
```diff
@@ -0,0 +1,128 @@
+                                          mc_path, FREQUENCY)
+        if new_revision == next_revision:
+            break
+        else:
```
No `else:` after `break`.

Pike commented on helper.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def switch_to_revision(rev, mc_path):
+    os.chdir(mc_path)
```
this should also use `--cwd` and not change the working dir?

zbraniecki commented:
Updated to feedback. Ready for review.

zbraniecki commented:
Thanks! I like it!

Pike commented on src/arewefluentyet/aggregate.py:
```diff
@@ -0,0 +1,128 @@
+
+    progress_data.append({
+        "data": progress,
+        "date": rev_date,
```
Use next_date instead of rev_date, and remove the get_revision_date code.
